### PR TITLE
Affiliates v2

### PR DESCRIPTION
**Pull request by santosinarp**
Affiliates V2 done with LockedSOV implementation.
So the rewards for referrer will be stored in LockedSOV contract.
For the deployment, can use the deployAffiliate() in contract_interaction.py

### CODE REVIEW

File: contracts/core/State.sol [hunk 1]
```diff
@@ -21,7 +22,8 @@ import "../interfaces/IWrbtcERC20.sol";
  * */
 contract State is Objects, ReentrancyGuard, Ownable {
 	using SafeMath for uint256;
-	using EnumerableBytes32Set for EnumerableBytes32Set.Bytes32Set;
+	using EnumerableAddressSet for EnumerableAddressSet.AddressSet; // enumerable map of addresses
```
Comment (tjcloa) [inline contracts/core/State.sol:25]:
```suggestion
	using EnumerableAddressSet for EnumerableAddressSet.AddressSet; // enumerable map of addresses
```
```suggestion
```
can be removed as not used and EnumerableBytes32Set has methods for addresses adding/removing

File: contracts/mixins/EnumerableAddressSet.sol [hunk 2]
```diff
@@ -0,0 +1,143 @@
+pragma solidity ^0.5.0;
```
Comment (tjcloa) [inline contracts/mixins/EnumerableAddressSet.sol:1]: file contracts/mixins/EnumerableAddressSet.sol can be removed as it is not used

File: contracts/core/State.sol [hunk 3]
```diff
@@ -6,6 +6,7 @@
 pragma solidity 0.5.17;
 
 import "./Objects.sol";
+import "../mixins/EnumerableAddressSet.sol";
```
Comment (tjcloa) [inline contracts/core/State.sol:9]:
```suggestion
import "../mixins/EnumerableAddressSet.sol";
```
```suggestion

```
can be removed

File: contracts/core/State.sol [hunk 4]
```diff
@@ -160,6 +162,18 @@ contract State is Objects, ReentrancyGuard, Ownable {
...
+	mapping(address => bool) public userNotFirstTradeFlag; // The flag is set on the user's first trade or borrowing
+
+	mapping(address => address) public affiliatesUserReferrer; // User => referrer (affiliate)
+	mapping(address => EnumerableAddressSet.AddressSet) internal referralsList; // list of referral addresses that owned by the referrer
+
+	uint256 public minReferralsToPayout = 3;
+	mapping(address => uint256) public affiliateRewardsHeld; // Total affiliate rewards that held in the protocol (Because the minimum referrals is less than the rule)
+	address public sovTokenAddress; // For affiliates SOV Bonus proccess
+	address public lockedSOVAddress;
+
```
Comment (tjcloa) [inline contracts/core/State.sol:176]: ⚠️ just to remind that before merging needs to be synced from the development branch and verified that all new variables are appended to avoid storage memory overlapping ⚠️

File: contracts/events/AffiliatesEvents.sol [hunk 5]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright 2017-2020, bZeroX, LLC. All Rights Reserved.
```
Comment (tjcloa) [inline contracts/events/AffiliatesEvents.sol:2]:
```suggestion
 * Copyright 2017-2020, bZeroX, LLC. All Rights Reserved.
```
```suggestion
 * Copyright 2021, Sovryn. All Rights Reserved.
```

File: contracts/events/FeesEvents.sol [hunk 6]
```diff
@@ -20,4 +20,12 @@ contract FeesEvents {
...
 
 	event EarnReward(address indexed receiver, address indexed token, bytes32 indexed loanId, uint256 amount);
+
+	/*event PayTradingFeeToAffiliate(
+		address indexed payer,
+		address indexed affiliateReferrer,
+		address indexed token,
+		bytes32 loanId,
+		uint256 amount
+	);*/
```
Comment (tjcloa) [inline contracts/events/FeesEvents.sol:30]:
```suggestion

	/*event PayTradingFeeToAffiliate(
		address indexed payer,
		address indexed affiliateReferrer,
		address indexed token,
		bytes32 loanId,
		uint256 amount
	);*/
```
```suggestion
```

File: contracts/interfaces/ISovryn.sol [hunk 7]
```diff
@@ -302,4 +314,20 @@ contract ISovryn is
...
+
+	function payTradingFeeToAffiliatesReferrer(
+		address referrer,
+		address token,
+		uint256 tradingFeeTokenBaseAmount
+	) external returns (uint256 referrerTradingFee);
+
+	function setAffiliatesUserReferrer(address user, address referrer) external; //onlyCallableByLoanPools
+
+	// function getAffiliatesUserReferrer(address user) external returns ; //AUDIT: do we need it to be public?
```
Comment (tjcloa) [inline contracts/interfaces/ISovryn.sol:332]:
```suggestion
	// function getAffiliatesUserReferrer(address user) external returns ; //AUDIT: do we need it to be public?
```
```suggestion
```

File: contracts/mockup/MockLockedSov.sol [hunk 8]
```diff
@@ -0,0 +1,17 @@
+pragma solidity 0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "../core/State.sol";
+
+contract LockedSovMockup is State {
```
Comment (tjcloa) [inline contracts/mockup/MockLockedSov.sol:6]: contract and file names if there is one contract should be the same
suggested to rename file and respective references

File: contracts/modules/Affiliates.sol [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+	//AUDIT Do we need to allow an owner to nullify a referrer for some reason?
+	/*
+    function affiliatesRemoveUserReferrer(address user, address referrer) external onlyOwner {
+        if(referrer != address(0) && getUserNotFirstTradeFlag[user]) {
+            delete affiliatesUserReferrer[user];
+            //TODO: event with the reason for removing referrer
+        }
+    }
+    */
```
Comment (tjcloa) [inline contracts/modules/Affiliates.sol:89]:
```suggestion
	//AUDIT Do we need to allow an owner to nullify a referrer for some reason?
	/*
    function affiliatesRemoveUserReferrer(address user, address referrer) external onlyOwner {
        if(referrer != address(0) && getUserNotFirstTradeFlag[user]) {
            delete affiliatesUserReferrer[user];
            //TODO: event with the reason for removing referrer
        }
    }
    */
```
```suggestion
```

File: contracts/modules/ProtocolSettings.sol [hunk 10]
```diff
@@ -67,6 +67,48 @@ contract ProtocolSettings is State, ProtocolTokenUser, ProtocolSettingsEvents {
...
+		require(Address.isContract(newLockedSOVAddress), "newLockSOVAddress not a contract");
+
+		address oldLockedSOVAddress = lockedSOVAddress;
+		lockedSOVAddress = newLockedSOVAddress;
+
+		emit SetLockedSOVAddress(msg.sender, oldLockedSOVAddress, newLockedSOVAddress);
+	}
+
+	function setMinReferralsToPayoutAffiliates(uint256 newMinReferrals) external onlyOwner {
+		require(newMinReferrals > 0, "Minimum referrals must be greater than 0");
```
Comment (tjcloa) [inline contracts/modules/ProtocolSettings.sol:106]: what if we will need to cancel this limitation?
```suggestion
		require(newMinReferrals > 0, "Minimum referrals must be greater than 0");
```
```suggestion
```

File: scripts/deployment/deploy_loanToken.py [hunk 11]
```diff
@@ -198,6 +200,102 @@ def testDeployment(acct, sovryn, loanTokenAddress, underlyingToken, collateralTo
     print("position size is ", collateral)
     tx = sovryn.closeWithSwap(loanId, acct, collateral, True, b'')
 
+'''
+test the affiliate margin trade by entering and closing a trade position
```
Comment (tjcloa) [inline scripts/deployment/deploy_loanToken.py:204]: why are tests here?
can integration testing be moved to hardhat?

**Comment by powerhousefrank** [comment]
The coveralls shows failure due to decreased coverage, saw these two files mainly contributing to the cause.

These are the files where the coverage is decreased (you can use the key `n` and `b` to go forward and backward of uncovered lines).
https://coveralls.io/builds/39878342/source?filename=contracts%2Fconnectors%2Floantoken%2FLoanTokenSettingsLowerAdmin.sol
https://coveralls.io/builds/39878342/source?filename=contracts%2Fmodules%2FLoanMaintenance.sol

Coveralls failure is more towards quality than test failure, if you could cover those edge cases as well, that would be nice.

**Comment by powerhousefrank** [comment]
https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/66d9d727d9ab645f5c10d0bf9bdc94f9047bc047/contracts/mockup/lockedSOVMockup.sol

Another merge with `development` might be needed.

**Comment by powerhousefrank** [comment]
https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/66d9d727d9ab645f5c10d0bf9bdc94f9047bc047/contracts/mockup/MockLockedSov.sol

This could be deleted.

**Comment by powerhousefrank** [comment]
https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/66d9d727d9ab645f5c10d0bf9bdc94f9047bc047/contracts/mockup/lockedSOVFailedMockup.sol

What is it's role?

**Comment by powerhousefrank** [comment]
https://github.com/DistributedCollective/Sovryn-smart-contracts/tree/development/contracts/locked

This is in `development` but newer changes are in `liquidity-mining` (which might be merged soon by James.)

File: contracts/core/State.sol (hunk 1 above)
Comment (santosinarp) [reply contracts/core/State.sol:25]: We still use the EnumerableAddressSet for referralsList storage

File: contracts/mockup/MockLockedSov.sol (hunk 8 above)
Comment (santosinarp) [reply contracts/mockup/MockLockedSov.sol:6]: This file is no longer used, i will remove

File: contracts/modules/ProtocolSettings.sol (hunk 10 above)
Comment (santosinarp) [reply contracts/modules/ProtocolSettings.sol:106]: If cancel the limitation is needed, then i will remove the require

File: scripts/deployment/deploy_loanToken.py (hunk 11 above)
Comment (santosinarp) [reply scripts/deployment/deploy_loanToken.py:204]: There is initialized loanToken test inside the deployment script. So i tried to reuse it.

**Comment by santosinarp** [comment]
> https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/66d9d727d9ab645f5c10d0bf9bdc94f9047bc047/contracts/mockup/lockedSOVFailedMockup.sol
> 
> What is it's role?



> https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/66d9d727d9ab645f5c10d0bf9bdc94f9047bc047/contracts/mockup/lockedSOVFailedMockup.sol
> 
> What is it's role?

It's used to simulate the failed process of depositSOV in lockedSOV contract. In case it's failed, the affiliates contract will log it.
This is just used for testing purposes

File: contracts/modules/Affiliates.sol [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+		bool success;
+		bool alreadySet;
+		bool userNotFirstTradeFlag;
+	}
+
+	function setAffiliatesReferrer(address user, address referrer) external onlyCallableByLoanPools {
+		SetAffiliatesReferrerResult memory result;
+
+		result.userNotFirstTradeFlag = getUserNotFirstTradeFlag(user);
+		result.alreadySet = affiliatesUserReferrer[user] != address(0) ? true : false;
```
Comment (eMarchenko) [inline contracts/modules/Affiliates.sol:68]: `result.alreadySet = affiliatesUserReferrer[user] != address(0);` is enough

File: contracts/modules/Affiliates.sol [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+    Module: Affiliates upgradable
+    Storage: from State, functions called from Protocol by delegatecall
+    */
+
+	using SafeERC20 for IERC20;
+
+	constructor() public {}
+
+	function() external {
+		revert("Affiliates - fallback not allowed");
```
Comment (eMarchenko) [inline contracts/modules/Affiliates.sol:27]: fallback with revert is unnecessary, unless you want custom revert message

File: contracts/modules/Affiliates.sol [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+		if (result.success) {
+			affiliatesUserReferrer[user] = referrer;
+			referralsList[referrer].add(user);
+			emit SetAffiliatesReferrer(user, referrer);
+		} else {
+			emit SetAffiliatesReferrerFail(user, referrer, result.alreadySet, result.userNotFirstTradeFlag);
+		}
+	}
+
+	function getReferralsList(address referrer) public view returns (address[] memory refList) {
```
Comment (eMarchenko) [inline contracts/modules/Affiliates.sol:79]: `external`

File: contracts/modules/Affiliates.sol [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+		address _priceFeeds = priceFeeds;
+
+		//calculate the reward amount, querying the price feed
+		(bool success, bytes memory data) =
+			_priceFeeds.staticcall(
+				abi.encodeWithSelector(
+					IPriceFeeds(_priceFeeds).queryReturn.selector,
+					feeToken,
+					sovTokenAddress, // dest token = SOV
+					feeAmount.mul(_getAffiliatesTradingFeePercentForSOV()).div(10**20)
```
Comment (eMarchenko) [inline contracts/modules/Affiliates.sol:129]: 10**20 -> 1e20

File: contracts/modules/Affiliates.sol [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+				)
+			);
+		assembly {
```
Comment (eMarchenko) [inline contracts/modules/Affiliates.sol:132]: is `abi.decode` available? it's a cleaner way to parse bytes

File: contracts/modules/Affiliates.sol [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+
+	/**
+	 * @param tradingFeeTokenBaseAmount total trading fee amount, the base for calculating referrer's fees
+	 */
+	function payTradingFeeToAffiliatesReferrer(
+		address referrer,
+		address token,
+		uint256 tradingFeeTokenBaseAmount
+	) external onlyCallableInternal returns (uint256 referrerBonusSovAmount, uint256 referrerBonusTokenAmount) {
+		bool isHeld = referralsList[referrer].length() >= getMinReferralsToPayout() ? false : true;
```
Comment (eMarchenko) [inline contracts/modules/Affiliates.sol:149]: `bool isHeld = referralsList[referrer].length() < getMinReferralsToPayout();`
in general, `true` and `false` are only required in single-word statements like `return true;`, `while(true)` or `someFlag = false;`

**Review by eMarchenko — APPROVED** [review]
only minor issues

**Comment by powerhousefrank** [comment]
Please test this on testnet first before merging with `development` branch. I have reset the development branch, you might have to create a new PR or revive this one @santosinarp 

Please review the testing on testnet @tjcloa

**Comment by santosinarp** [comment]
Need to re-open this PR because lack of testing on the testnet

**Comment by santosinarp** [comment]
Need to create new PR for this feature due to lack of testing on testnet.
The new PR is created at https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/280
